nikonov.andrew/S1

ysffmn opened the pull request:
FWD + RING + FAKE
![S1 критерии](https://github.com/user-attachments/assets/894a50ff-76bf-46e5-a948-d73837d2950b)

ysffmn committed:
initial commit

ysffmn committed:
first ver. of input func

ysffmn committed:
input / output func ver.2

ysffmn committed:
list header added

ysffmn committed:
technical commit

ysffmn committed:
first iterator impl

ysffmn committed:
Merge branch 'master' into nikonov.andrew/S1

ysffmn committed:
quick cg fix

ysffmn committed:
build qF

ysffmn committed:
overflow check added

ysffmn committed:
attempt to solve overflow error

ysffmn committed:
same as prev

ysffmn committed:
same as prev

ysffmn committed:
leak has been eliminated

ysffmn committed:
brain rot: the beginning

ysffmn committed:
it should workes, but it doesn't

ysffmn committed:
int -> size_t; iterator fixed;

ysffmn committed:
i guess the main problem isn't in the code, but it's sitting in front of the monitor.

ysffmn committed:
some changes in output func

ysffmn committed:
cg

ysffmn committed:
List is in progress, but code is mega shit

ysffmn committed:
List in progress 2

ysffmn committed:
cg

ysffmn committed:
List in progress 3

ysffmn committed:
technical commit

ysffmn committed:
List in progress 4

ysffmn committed:
List in progress 5 and Test headers added

Volgarenok commented:
> FWD + RING + FAKE ![image](https://private-user-images.githubusercontent.com/174838756/418311536-2225e2e9-0b8b-48bd-a3e3-f04c07b15bca.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEyNzU1OTEsIm5iZiI6MTc0MTI3NTI5MSwicGF0aCI6Ii8xNzQ4Mzg3NTYvNDE4MzExNTM2LTIyMjVlMmU5LTBiOGItNDhiZC1hM2UzLWYwNGMwN2IxNWJjYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMwNlQxNTM0NTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00YWNiN2U0NjJkYTZhZmE3NGE1ODg5ZDcyMTlhNjAwYWU2ZmM3OTliNDdiOTk0MjBjOWQ0MDU3YWFlYTgyMzViJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ss0hdQ-DyexBSISukUAmDMzHfVrJhwWMyEueWSazJo4)

Sign 1: Не приложили таблицу с критериями и выбранным баллом

ysffmn committed:
fixed memory allocation for fake node, made some fixes and file splitting

ysffmn committed:
added const iterator

ysffmn committed:
typo fix

ysffmn committed:
some iterator and list fiexes, new constructors added

ysffmn committed:
a few new constructors and some method fixes

ysffmn committed:
op=, erase and insert added, but it looks not really good. i think my whole code is shit..

ysffmn committed:
size() fix

ysffmn committed:
complited assign func

ysffmn committed:
it was obvious

ysffmn committed:
transfer commit (prev. - technical commit)

ysffmn committed:
i'm fucked up, terminal is going to hell after running boost test

ysffmn committed:
some critical fixes

ysffmn committed:
finnaly I found that sneaky mistake

ysffmn committed:
constructors test done

ysffmn committed:
added tests..

ysffmn committed:
replased standard list

ysffmn committed:
fixed empty conditions

ysffmn committed:
added inheritance

ysffmn committed:
added operators overload, splice functions, changed list costructor, RESTRUCTURISATION included

ysffmn committed:
cg fixes

ysffmn committed:
sfinae deleted

ysffmn committed:
added op tests and lots of fixes(iters, compare func, typos, etc)

ysffmn committed:
some tests added and splise fixes

ysffmn committed:
noexcept added

ysffmn committed:
added range erase and littele template naming changes

ysffmn committed:
added overloaded insert and some little fixes

ysffmn committed:
whitespace fix

ysffmn committed:
labda executed

ysffmn committed:
ListNode -> detail::

ysffmn committed:
whitespaces fix

ysffmn committed:
typo fix

Volgarenok commented on nikonov.andrew/S1/List.hpp:
```diff
@@ -0,0 +1,727 @@
+  template< typename T >
+  class List;
+}
+namespace detail
```
Пространство имён detail должно быть вложено в ваше

Volgarenok commented on nikonov.andrew/S1/List.hpp:
```diff
@@ -0,0 +1,727 @@
+#include "ListIterator.hpp"
+#include <cstddef>
+#include <cstdlib>
+#include <initializer_list>
```
CG

Volgarenok commented on nikonov.andrew/S1/List.hpp:
```diff
@@ -0,0 +1,727 @@
+#define LIST_HPP
+#include "ListIterator.hpp"
+#include <cstddef>
+#include <cstdlib>
```
cstdlib зачем?..

Volgarenok commented on nikonov.andrew/S1/List.hpp:
```diff
@@ -0,0 +1,727 @@
+  template< typename T >
+  bool tLess(const T & a, const T & b);
+  template< typename T >
+  bool tGreater(const T & a, const T & b);
```
Что значит `t`? CG у нас какой?

Volgarenok commented on nikonov.andrew/S1/List.hpp:
```diff
@@ -0,0 +1,727 @@
+  template< typename T >
+  using const_iterator = ConstListIterator< T >;
+  template< typename T >
+  using node_t = detail::ListNode< T >;
```
А для деревьев вы итераторы как будете добавлять? У вас конфликт имён будет

Volgarenok commented on nikonov.andrew/S1/List.hpp:
```diff
@@ -0,0 +1,727 @@
+    node_t< T > * fake;
+  public:
+    List();
+    List(const List & copy);
+    List(List && copy);
```
А с операторами присваивания что?

Volgarenok commented on nikonov.andrew/S1/List.hpp:
```diff
@@ -0,0 +1,727 @@
+
+  template< typename T >
+  List< T >::List(const List< T > & copy):
+    fake(static_cast< node_t< T > * >(malloc(sizeof(node_t< T >))))
```
Почему `malloc`?

Volgarenok commented on nikonov.andrew/S1/List.hpp:
```diff
@@ -0,0 +1,727 @@
+
+  template< typename T >
+  List< T >::List(size_t k, const T & value):
+    fake(static_cast< node_t< T > * >(malloc(sizeof(node_t< T >))))
```
Желательно подумать, как сделать так, чтоб не приходилось явно создание фейкового узла везде писать. Всё-таки это логика нетривиальная, а вы её копируете везде

Volgarenok commented on nikonov.andrew/S1/List.hpp:
```diff
@@ -0,0 +1,727 @@
+    node_t< T > * tailNode = nullptr;
+    for (size_t i = 0; i < k; ++i)
+    {
+      node_t< T > * newNode = new node_t< T >{ value, fake };
```
Утечка получается, если тут исключение

Volgarenok commented on nikonov.andrew/S1/List.hpp:
```diff
@@ -0,0 +1,727 @@
+  {
+    fake->next = fake;
+    node_t< T > * curr = fake;
+    for (auto el : il)
```
CG: обратите внимание на требования к расстановке пробелов у двоеточия

Volgarenok commented on nikonov.andrew/S1/List.hpp:
```diff
@@ -0,0 +1,727 @@
+    node_t< T > * curr = fake;
+    for (auto el : il)
+    {
+      node_t< T > * newNode = new node_t< T >{ el, fake };
```
Тоже утечка

Volgarenok commented on nikonov.andrew/S1/List.hpp:
```diff
@@ -0,0 +1,727 @@
+      node_t< T > * newNode = new node_t< T >{ el, fake };
+      curr->next = newNode;
+      curr = newNode;
+    }
+  }
+
+  template< typename T >
+  template< typename InputIterator >
+  List< T >::List(InputIterator begin, InputIterator end):
+    fake(static_cast< node_t< T > * >(malloc(sizeof(node_t< T >))))
+  {
+    fake->next = fake;
+    node_t< T > * curr = fake;
+    for (; begin != end; ++begin)
+    {
+      node_t< T > * newNode = new node_t< T >{ *begin, fake };
+      curr->next = newNode;
+      curr = newNode;
+    }
+  }
```
Ну тут всё небезопасно. Исправляйте

Volgarenok commented on nikonov.andrew/S1/List.hpp:
```diff
@@ -0,0 +1,727 @@
+    {
+      return true;
+    }
+    return detail::compareLists(*this, rhs, detail::tGreater< T >);
```
Всё-таки операторы друг через друга нужно реализовывать

Volgarenok commented on nikonov.andrew/S1/List.hpp:
```diff
@@ -0,0 +1,727 @@
+  {
+    T * copy(new T(value));
+    push_front(*copy);
+  }
```
Чё происходит вообще?.. А где delete для этого объекта? У вас утечка в чистом виде. Вы тестировали этот метод?

Volgarenok commented on nikonov.andrew/S1/List.hpp:
```diff
@@ -0,0 +1,727 @@
+  }
+
+  template< typename T >
+  void List< T >::splice(const_iterator< T > pos, List< T > && x, const_iterator< T > first, const_iterator< T > last) noexcept
```
Очень много дублирований в splice-ах. Подумайте, какой из них самый общий. Или может нужно какой-то splice более общий сделать, чтоб через него реализовать остальные?

Volgarenok commented on nikonov.andrew/S1/List.hpp:
```diff
@@ -0,0 +1,727 @@
+    return ++curr;
+  }
+
+  template< typename T >
+  iterator< T > List< T >::insert(const_iterator< T > position, T && val)
```
А в чём смысл этих метод с r-value, если вы по итогу вызываете те, что с l-value? Могли бы вообще ничего не писать: ссылка на константу отлично связывается с r-value и всё то же самое было бы

Что происходить-то должно? Работа отдельно с R-value нам для чего нужна?

Volgarenok commented on nikonov.andrew/S1/List.hpp:
```diff
@@ -0,0 +1,727 @@
+  bool tGreater(const T & a, const T & b)
+  {
+    return a > b;
+  }
```
Не надо это писать. Используйте стандартные

Volgarenok commented on nikonov.andrew/S1/ListIterator.hpp:
```diff
@@ -0,0 +1,166 @@
+  {
+    detail::ListNode< T > * node;
+    using this_t = ListIterator< T >;
+
```
А с инкапсуляцией что?

Volgarenok commented on nikonov.andrew/S1/ListIterator.hpp:
```diff
@@ -0,0 +1,166 @@
+    using this_t = ListIterator< T >;
+
+    ListIterator();
+    explicit ListIterator(detail::ListNode< T > * ptr = nullptr);
```
Этот конструктор - деталь реализации

Volgarenok commented on nikonov.andrew/S1/additional-utilities.cpp:
```diff
@@ -0,0 +1,152 @@
+nikonov::List< std::pair< std::string, nikonov::List< size_t > > > * nikonov::getPList(std::istream& in, bool& overflowFlag)
+{
+  using pair_t = std::pair< std::string, List< size_t > >;
+  List< pair_t > * pList = new List< pair_t >;
```
А сами листы зачем складывать в динамическую память?

Volgarenok commented on nikonov.andrew/S1/additional-utilities.cpp:
```diff
@@ -0,0 +1,152 @@
+  }
+  return pList;
+}
+void nikonov::processPList(List< std::pair< std::string, List< size_t > > > * pList, bool& overflowFlag)
```
Никаких указателей не должно быть в клиентском коде в принципе. Мы для этого структуру и писали, чтоб работу с памятью инкапсулировать в контейнер

ysffmn committed:
some fixes, still got work to do

ysffmn committed:
List: added assignment ops
Iterator: fixed incapsulation issue
add-utils: getIterAt const denial issue fixed

ysffmn committed:
Iterator: incapsulation fix

ysffmn committed:
splice fixes, memore leak fixes, few name changes

ysffmn committed:
wp fix

ysffmn committed:
constructor fix

ysffmn committed:
im done, i gonna be a pluber

ysffmn committed:
test fix

ysffmn committed:
List fix

ysffmn committed:
i have no idea how to fix it another way

ysffmn committed:
simplified getPList
